pytorch_lightning/loops/training_batch_loop.py

```diff
@@ -0,0 +1,722 @@
+                    self.warning_cache.warn(
+                        "`training_step` hook signature has changed in v1.3."
+                        " `optimizer_idx` argument has been removed in case of manual optimization. Support for"
+                        " the old signature will be removed in v1.5", DeprecationWarning
+                    )
+                args.append(opt_idx)
+            elif not self.trainer.has_arg(
+                "training_step", "optimizer_idx"
+            ) and self.trainer.lightning_module.automatic_optimization:
+                raise ValueError(
+                    f"Your LightningModule defines {len(self.trainer.optimizers)} optimizers but"
+                    ' `training_step` is missing the `optimizer_idx` argument.'
+                )
+
+        # pass hiddens if using tbptt
+        if self.trainer.truncated_bptt_steps is not None:
+            args.append(hiddens)
+
+        return args
+
+    def run_optimization_start(self, opt_idx: int, optimizer: torch.optim.Optimizer) -> None:
+        """Toggles the optimizer to ensure the correct one is used and prevend dangling grads.
+
+        Args:
+            opt_idx: the index of the optimizer to use
+            optimizer: the optimizer to use
+
+        """
+        # make sure only the gradients of the current optimizer's parameters are calculated
+        # in the training step to prevent dangling gradients in multiple-optimizer setup.
+        if self.trainer.lightning_module.automatic_optimization and len(self.trainer.optimizers) > 1:
+            model = self.trainer.lightning_module
+            model.toggle_optimizer(optimizer, opt_idx)
+
+    @contextmanager
+    def block_ddp_sync_behaviour(self, should_block_sync: bool = False) -> Generator[None, None, None]:
+        """
+        automatic_optimization = True
+        Blocks ddp sync gradients behaviour on backwards pass.
+        This is useful for skipping sync when accumulating gradients, reducing communication overhead
+
+        automatic_optimization = False
+        do not block ddp gradient sync when using manual optimization
+        as gradients are needed within the training step
+
+        Returns:
+            context manager with sync behaviour off
+
+        """
+        if (
+            isinstance(self.trainer.training_type_plugin, ParallelPlugin)
+            and (self.trainer.lightning_module.automatic_optimization or should_block_sync)
+        ):
+            with self.trainer.training_type_plugin.block_backward_sync():
+                yield None
+        else:
+            yield None
+
+    def training_step_and_backward(
+        self, split_batch: Any, batch_idx: int, opt_idx: int, optimizer: torch.optim.Optimizer,
+        hiddens: Optional[Tensor]
+    ) -> STEP_OUTPUT:
+        """Wrap forward, zero_grad and backward in a closure so second order methods work"""
+        with self.trainer.profiler.profile("training_step_and_backward"):
```

Comment:
thinking about using a wrapper in case that all function body is in with block

Reply:
the profiling context here is tied to an object, I don't think we can make a context manager so easily. And besides, I think it is safer so we can make sure we profile exactly only the operations under the indented block and not accidentally the whole method.